runtime/decimal.c

```diff
@@ -3,49 +3,69 @@
 
 typedef GC decQuad *DecimalPtr;
 
-typedef struct {
-    TaggedPtr ptr;
-    PanicCode panicCode;
-} TaggedPtrPanicCode;
-
 // Status flags that we should treat as failure
 // XXX need to investigate DEC_Underflow
 #define DECIMAL_STATUS_FAIL (DEC_Errors|DEC_Subnormal)
 
-static inline PanicCode statusPanicCode(uint32_t status) {
-    if (status & DECIMAL_STATUS_FAIL) {
-        // XXX decide what panic codes to use 
-        return (status & DEC_Division_by_zero) ? PANIC_DIVIDE_BY_ZERO : PANIC_ARITHMETIC_OVERFLOW;
-    }
-    return 0;
-}
+static TaggedPtrPanicCode finish(decQuad *dq, decContext *cx);
 
 static inline TaggedPtr createDecimal(decQuad *dq) {
     DecimalPtr dp = _bal_alloc(sizeof(decQuad));
     *dp = *dq;
-    return ptrAddFlags(dp, (uint64_t)TAG_DECIMAL);
+    return ptrAddFlags(dp, (uint64_t)TAG_DECIMAL << TAG_SHIFT);
 }
 
-static inline decQuad *taggedToDecQuad(TaggedPtr tp) {
+static inline const decQuad *taggedToDecQuad(TaggedPtr tp) {
     UntypedPtr p = taggedToPtr(tp);
     return (void *)p;
 }
 
+static inline void initContext(decContext *cx) {
+    cx->status = 0;
+    cx->round = DEC_ROUND_HALF_EVEN;
+}
+
 TaggedPtrPanicCode _bal_decimal_add(TaggedPtr tp1, TaggedPtr tp2) {
-    TaggedPtrPanicCode result;
-    decQuad *dq1 = taggedToDecQuad(tp1);
-    decQuad *dq2 = taggedToDecQuad(tp2);
+    const decQuad *dq1 = taggedToDecQuad(tp1);
+    const decQuad *dq2 = taggedToDecQuad(tp2);
     decQuad d;
     decContext cx;
-    cx.status = 0;
-    cx.round = DEC_ROUND_HALF_EVEN;
+    initContext(&cx);
     decQuadAdd(&d, dq1, dq2, &cx);
-    result.panicCode = statusPanicCode(cx.status);
-    if (result.panicCode == 0) {
-        result.ptr = createDecimal(&d);
-    }
-    return result;    
+    return finish(&d, &cx);    
 }
 
+TaggedPtrPanicCode _bal_decimal_sub(TaggedPtr tp1, TaggedPtr tp2) {
+    const decQuad *dq1 = taggedToDecQuad(tp1);
+    const decQuad *dq2 = taggedToDecQuad(tp2);
+    decQuad d;
+    decContext cx;
+    initContext(&cx);
+    decQuadSubtract(&d, dq1, dq2, &cx);
```

Comment:
The dq1 and dq2 variables aren't buying us anything. So let's inline them.

```
decQuadSubtract(&d, taggedToDecQuad(tp1), taggedToDecQuad(tp2), &cx);
```

You can do this in the next pull request.